### ISSUE TITLE
Adhere to max-weight limits within LCP state machine

### ISSUE THREAD

**Issue by Roasbeef** [lightningnetwork/lnd#28]
Currently within the [commitment update state-machine](https://github.com/lightningnetwork/lnd/blob/master/lnwallet/channel.go) all HTLC adds are accepted. This is an issue since the addition of a new HTLC may push the commitment transaction over the consensus enforced weight (formerly called cost) limit for transactions (4,000,000). An additional factor to be considered is the weight of a future stealing transaction that sweeps all the HTLC's within a invalidly broadcast commitment transaction. 

The LCP state machine should be modified to _reject_ adding an HTLC if it puts both the commitment transaction, and a future stealing transaction over the current weight limit.

**Comment by AndrewSamokhvalov** [comment]
@Roasbeef Could you elaborate please what "future stealing transaction" mean?

**Comment by Roasbeef** [comment]
By "stealing transaction" I mean the transaction(s) that sweep any rightfully owned outputs into the node's wallet after a unilateral/uncooperative channel closure directly on-chain. In this scenario there may be zero (no HTLC's, they broadcasted) or many outputs (fully loaded commitment tx with many HTLC's) to sweep depending on who broadcasted the commitment transaction, and if it's the most up-to-date version or not. 

The goal of this ticket is to ensure that the commitment transaction is in a broadcast-able state _at all times_. This entails observation of a number of constraints including proper fee rate, the absence of any dust outputs, and finally the total weight of the commitment transaction (which this ticket aims to address). 

There are two ways to go about this, both with their own tradeoffs: 

1. Ensure that the weight of the commitment transactions stays below 4,000,000 (the consensus limit) at all times. Given that we want to ensure through propagation of our transactions, we shuold likely instead target the 400,000 weight limit enforced by policy within most implementations. This would entail modifying channel state machine slightly to keep tally of the current pending weight of the commitment transaction if a new state transition were to immediately be executed. This check can either be placed in `AddHTLC` or within `SignNextCommitment`.  that all peers adhere to this limit, then this should only possible when adding new HTLC's created to fulfill a local payment are encountered. An error should be returned to the caller, either immediately, or after a few attempts (would still be added to the batch). 
2. Ensure that the weight of the commitment transaction stays below 4,000,000 (or 400,000) _AND_ the weight of the transaction attempting to sweep _all_ outputs is also below 4,000,000 (or 400,000). This calculation will need to factor in the size of the witnesses within all inputs of the sweeping transaction. 

On one hand, the first option allows us to have a greater number of pending HTLC's on a commitment transaction, increasing the available multiplexed payment bandwidth a node has. However, sweeps (in the very worst-case) may have to be carried out with one or more transactions if the weight of the initial transaction that attempts to sweep them all is too large. 

Similarly, the second option limits the avilable multiplexed payment bandwidth, but ensure that any sort of dispute can be _atomically_ resolved with (potentially a single transaction). 

One other aspect to consider is that as pending HTLC's, and the delay commitment output will likely have heterogeneous timeouts/delays. As a result, it may not actually be possible in practice to create a single transaction which sweeps all outputs as the timeouts/delays will differ. Therefore, I think we should just target the first option, since it's simpler and doesn't relay on a possible simplistic view of the distribution of timeouts.

**Comment by AndrewSamokhvalov** [comment]
@Roasbeef Wow! Thanks for such detailed response!

**Comment by AndrewSamokhvalov** [comment]
* As far as I understand, correct me if I am wrong, the weight of the final transaction consist of  two parts:
* Initial commitment transaction (`OurCommitTx`) is generated during channel creation, and its size isn't  changing during life of the channel, so we may calculate size once and then use as base size, lets call it `initialCommitmentSize`.
* The size of the unsettle `htlc` multiplied by its count.

As far as the `htlc` is `TxOut` and the size of `TxOut` determined as:
```
// SerializeSize returns the number of bytes it would take to serialize the
// the transaction output.
func (t *TxOut) SerializeSize() int {
	// Value 8 bytes + serialized varint size for the length of PkScript +
	// PkScript bytes.
	return 8 + VarIntSerializeSize(uint64(len(t.PkScript))) + len(t.PkScript)
}
```
we should know the the of `PkScript` size in advance, but we can't, so then lets take the maximum size  which is `8 + 8(uint64) + 1(OP_0) + 32(SHA_256) = 49 = maxHtlcSize`.

So we should check the size `initialCommitmentSize + maxHtlcSize * len(GetUnsettledHTLCs())` to be not greater than `400000` bytes.

* Size `400000` bytes is almost half of the block size (`1000000`), shouldn't we consider attack when `Alice` sends `Bob` a lots of `htlcs` and potentially make his transaction not broadcast-able due to lack of fee?**

**Comment by Roasbeef** [comment]
Closed by #46.
